src/app/CommandHandler.cpp

```diff
@@ -284,33 +341,62 @@ void CommandHandler::DecrementHoldOff()
         }
         else if (!IsGroupRequest())
         {
-            CHIP_ERROR err = SendCommandResponse();
+            CHIP_ERROR err = InitiateSendingCommandResponses();
             if (err != CHIP_NO_ERROR)
             {
                 ChipLogError(DataManagement, "Failed to send command response: %" CHIP_ERROR_FORMAT, err.Format());
             }
         }
     }
 
+    if (mState == State::AwaitingResponse)
+    {
+        // If we are AwaitingResponse from command that was successfully sent out, the responsibility to call Close()
+        // is now expected to be completed by the ExchangeDelegate callbacks of CommandHandler. In order to get those
+        // callbacks we have to set ourselves as the delegate and return.
+        mExchangeCtx->SetDelegate(this);
+        return;
+    }
     Close();
 }
 
 CHIP_ERROR CommandHandler::SendCommandResponse()
 {
-    System::PacketBufferHandle commandPacket;
+    VerifyOrReturnError(!mChunks.IsNull(), CHIP_ERROR_INCORRECT_STATE);
+    System::PacketBufferHandle packet = mChunks.PopHead();
 
-    VerifyOrReturnError(mPendingWork == 0, CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrReturnError(mState == State::AddedCommand, CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrReturnError(mExchangeCtx, CHIP_ERROR_INCORRECT_STATE);
+    bool moreToSend               = !mChunks.IsNull();
+    Messaging::SendFlags sendFlag = Messaging::SendMessageFlags::kNone;
+    if (moreToSend)
+    {
+        sendFlag = Messaging::SendMessageFlags::kExpectResponse;
+    }
 
-    ReturnErrorOnFailure(Finalize(commandPacket));
-    ReturnErrorOnFailure(
-        mExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::InvokeCommandResponse, std::move(commandPacket)));
-    // The ExchangeContext is automatically freed here, and it makes mpExchangeCtx be temporarily dangling, but in
-    // all cases, we are going to call Close immediately after this function, which nulls out mpExchangeCtx.
+    // If this is the first response we are sending, and we are waiting on response to send out another
+    // InvokeResponseMessages we setup the response timer.
+    if (mState == State::AddedCommand && moreToSend)
+    {
+        mExchangeCtx->UseSuggestedResponseTimeout(app::kExpectedIMProcessingTime);
+    }
 
+    ReturnErrorOnFailure(
+        mExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::InvokeCommandResponse, std::move(packet), sendFlag));
     MoveToState(State::CommandSent);
+    if (moreToSend)
+    {
+        MoveToState(State::AwaitingResponse);
+    }
```

Comment:
We should not see 2 MoveToState in a row, because on exit of this method, the state is AwaitingResponse. Is there a side-effect to the first MoveToState?